### DISCUSSION TITLE
Reinforced rubber under a simple tensile test

### DISCUSSION THREAD

**Discussion by ZAARAOUI999** [adtzlr/felupe#318] [Show and tell]
Hello, Mr. Andreas, I wish you were well, so as is mentioned by the title of this discussion, I'm trying to simulate a reinforced rubber under a simple tensile test. So, the flow of my work is:
- Creating a rectangular geometry (20x20) containing randomly distributed inclusions (fibers)
- Extracting of the regions (so we will have a region for the matrix and a region for each inclusion) 
- Creating Fields (mixed field with n=1 (no volume change, no pressure on boundaries will be taken into consideration))
- Linking all the inclusions's fields to the matrix field
- Creating materials and solidbodies (MatrixSolid (NeoHook: C10=0.7 and k=2.0 (with plane strain flag from Matadi)) and five InclusionSolid (Linear Elastic Plane Strain: E=2000.0 and nu=0.3))
- Applying uniaxial test boundaries on the matrix 
- Setting the load (displacement of 10 mm)
- Creating a step and fire up the job
Here are the regions and the mesh
![regions](https://user-images.githubusercontent.com/115699524/198846838-67a63a70-a4e2-4c58-a631-b61b54acdff7.png)
![mesh](https://user-images.githubusercontent.com/115699524/198846841-4d71145b-28a2-4538-a394-9a94e4908597.png)

Then, I got this!!!
![composite](https://user-images.githubusercontent.com/115699524/198846623-4903e15a-5b27-48b7-b102-38daa2dc1ae1.gif)
![inclusions_LE_E_2000_NU_0_3](https://user-images.githubusercontent.com/115699524/198846678-4453b7c2-14bc-4101-9772-776213f13dae.gif)
![matrix_NH_C10_0_7_K_2](https://user-images.githubusercontent.com/115699524/198846723-220e7b48-5f40-4320-9c50-6597059ed4da.gif)

It looks like I have misunderstood things or this simulation needs the definition of a relation between fibers and matrix I mean at the level of the interface. And it's clear that the behavior of the fibers isn't normal. Any idea about what's happening here? 
Thanks in advance

**Comment by adtzlr** [comment]
Hi, thats indeed a bit tricky. Here is an example. Hope that helps.

```python
import felupe as fem
import numpy as np

# create a rectangle
m = fem.Rectangle(n=21)

# take some points from the inside for the fiber-reinforced area
eps = 1e-3
mask = np.arange(m.npoints)[np.logical_and.reduce([
    m.points[:, 0] >= 0.3,
    m.points[:, 0] <= 0.7 + eps,
    m.points[:, 1] >= 0.3,
    m.points[:, 1] <= 0.7 + eps,
])]

# copies of the mesh
mesh = [m.copy(), m.copy()]

# create sub-meshes (fiber, matrix)
mesh[0].update(m.cells[ np.all(np.isin(m.cells, mask), axis=1)])
mesh[1].update(m.cells[~np.all(np.isin(m.cells, mask), axis=1)])

# a global and two sub-regions
region = fem.RegionQuad(m)
regions = [fem.RegionQuad(me) for me in mesh]

# a global and two sub-fields
field = fem.FieldsMixed(region, n=1, planestrain=True)
fields = [
    fem.FieldsMixed(regions[0], n=1, planestrain=True),
    fem.FieldsMixed(regions[1], n=1, planestrain=True),
]

# two material model formulations
neo_hooke = fem.NeoHooke(mu=1, bulk=1)
linear_elastic = fem.LinearElastic(E=100, nu=0.3)

# the solid bodies
fiber = fem.SolidBody(umat=linear_elastic, field=fields[0])
matrix = fem.SolidBody(umat=neo_hooke, field=fields[1])

# boundary conditions; applied on the global field
# which must be the x0-argument in job.evaluate()
boundaries = dict(
    fixed=fem.Boundary(field[0], fx=0),
    move=fem.Boundary(field[0], fx=1),
)

# prepare a step with substeps
move = fem.math.linsteps([0, 0.5], num=10)
step = fem.Step(
    items=[matrix, fiber],
    ramp={boundaries["move"]: move}, 
    boundaries=boundaries
)

# take care of the x0-argument
job = fem.Job(steps=[step])
job.evaluate(x0=field, filename="result.xdmf")
```

Have a nice evening,
Andreas

![image](https://user-images.githubusercontent.com/5793153/198888990-2d6ecf72-893b-4b34-a9a6-bdc73bc74159.png)

**Reply by adtzlr** [reply]
This code with a fix for `job.evaluate(x0=...)` needs the latest git version (not released). I'll publish a new release in 1-2 weeks. For technical  details see #320.

**Reply by adtzlr** [reply]
<details>
<summary>This also works with plane-strain materials from matadi:</summary>

```python
import felupe as fem
import numpy as np

# create a rectangle
m = fem.Rectangle(n=21)

# take some points from the inside for the fiber-reinforced area
eps = 1e-3
mask = np.arange(m.npoints)[np.logical_and.reduce([
    m.points[:, 0] >= 0.3,
    m.points[:, 0] <= 0.7 + eps,
    m.points[:, 1] >= 0.3,
    m.points[:, 1] <= 0.7 + eps,
])]

# copies of the mesh
mesh = [m.copy(), m.copy()]

# create sub-meshes (fiber, matrix)
mesh[0].update(m.cells[ np.all(np.isin(m.cells, mask), axis=1)])
mesh[1].update(m.cells[~np.all(np.isin(m.cells, mask), axis=1)])

# a global and two sub-regions
region = fem.RegionQuad(m)
regions = [fem.RegionQuad(me) for me in mesh]

# a global and two sub-fields
field = fem.FieldsMixed(region, n=1)
fields = [
    fem.FieldsMixed(regions[0], n=1),
    fem.FieldsMixed(regions[1], n=1),
]

# two material model formulations
import matadi as mat

neo_hooke = mat.MaterialHyperelasticPlaneStrain(mat.models.neo_hooke, C10=0.5, bulk=1)
linear_elastic = fem.LinearElasticPlaneStrain(E=100, nu=0.3)

# the solid bodies
fiber = fem.SolidBody(umat=linear_elastic, field=fields[0])
matrix = fem.SolidBody(umat=neo_hooke, field=fields[1])

# boundary conditions; applied on the global field
# which must be the x0-argument in job.evaluate()
boundaries = dict(
    fixed=fem.Boundary(field[0], fx=0),
    move=fem.Boundary(field[0], fx=1),
)

# prepare a step with substeps
move = fem.math.linsteps([0, 0.5], num=10)
step = fem.Step(
    items=[matrix, fiber],
    ramp={boundaries["move"]: move}, 
    boundaries=boundaries
)

# take care of the x0-argument
job = fem.Job(steps=[step])
job.evaluate(x0=field, filename="result.xdmf")
```
</details>

**Reply by ZAARAOUI999** [reply]
That's worked, I'm very thankful for you, Thanks.

**Reply by adtzlr** [reply]
I'll update the docs when I have some free time. These parts are missing (define a global field for sub-fields) because I'm still trying to enhance the API - but this is a long-term project.

**Reply by ZAARAOUI999** [reply]
My results after your help, that's great!
![fantastic_felupe](https://user-images.githubusercontent.com/115699524/199117899-b0af4de7-413d-40e8-998c-3a0cba255b49.gif)

**Reply by adtzlr** [reply]
Congratulations 👏🏻 , finally it looks like a realistic strain distribution!
